pkg/pool-hooks/contracts/hooks-quantamm/HyperSurgeHook.sol

```diff
@@ -0,0 +1,747 @@
+    using FixedPoint for uint256;
+    using SafeCast for uint256;
+
+    error InvalidArrayLengths();
+    error TokenIndexOutOfRange();
+    error NumTokensOutOfRange();
+    error InvalidPairIndex();
+    error PoolNotInitialized();
+    error InvalidDecimals();
+    error InvalidSurgeFeePercentage();
+    error InvalidThresholdDeviation();
+    error InvalidCapDeviationPercentage();
+
+    struct TokenPriceCfg {
+        uint32 pairIndex;
+        uint32 priceDivisor;
+    }
+
+    struct PoolDetails {
+        uint32 arbMaxSurgeFeePercentage;
+        uint32 arbThresholdPercentage;
+        uint32 arbCapDeviationPercentage;
+        uint32 noiseMaxSurgeFeePercentage;
+        uint32 noiseThresholdPercentage;
+        uint32 noiseCapDeviationPercentage;
+        uint8 numTokens;
+        bool initialized;
+    }
+
+    struct PoolCfg {
+        PoolDetails details;
+        TokenPriceCfg[8] tokenCfg;
+    }
+
+    mapping(address => PoolCfg) private _poolCfg;
+
+    uint256 private immutable _defaultMaxSurgeFee;
+
+    uint256 private immutable _defaultThreshold;
+
+    uint256 private immutable _defaultCapDeviation;
+
+    constructor(
+        IVault vault,
+        uint256 defaultMaxSurgeFeePercentage,
+        uint256 defaultThresholdPercentage,
+        uint256 defaultCapDeviation,
+        string memory version
+    ) SingletonAuthentication(vault) VaultGuard(vault) Version(version) {
+        _ensureValidPct(defaultMaxSurgeFeePercentage);
+        _ensureValidPct(defaultThresholdPercentage);
+        _ensureValidPct(defaultCapDeviation);
+        _defaultMaxSurgeFee = defaultMaxSurgeFeePercentage;
+        _defaultThreshold = defaultThresholdPercentage;
+        _defaultCapDeviation = defaultCapDeviation;
+    }
+
+    ///@inheritdoc IHooks
+    function getHookFlags() external pure override returns (HookFlags memory hookFlags) {
+        hookFlags.shouldCallComputeDynamicSwapFee = true;
+        hookFlags.shouldCallAfterAddLiquidity = true;
+        hookFlags.shouldCallAfterRemoveLiquidity = true;
+    }
+
+    /// @inheritdoc IHooks
+    function onRegister(
+        address,
+        address pool,
+        TokenConfig[] memory tokenCfgs,
+        LiquidityManagement calldata
+    ) external override onlyVault returns (bool) {
+        PoolDetails memory details;
+        if (tokenCfgs.length >= 2 && tokenCfgs.length <= 8) {
+            details.arbMaxSurgeFeePercentage = _defaultMaxSurgeFee.toUint32();
+            details.arbThresholdPercentage = _defaultThreshold.toUint32();
+            details.arbCapDeviationPercentage = _defaultCapDeviation.toUint32();
+            details.noiseMaxSurgeFeePercentage = _defaultMaxSurgeFee.toUint32();
+            details.noiseThresholdPercentage = _defaultThreshold.toUint32();
+            details.noiseCapDeviationPercentage = _defaultCapDeviation.toUint32();
+
+            details.numTokens = uint8(tokenCfgs.length);
+
+            details.initialized = true;
```

Comment:
Maybe the bool `initialized` is not needed. 

I understand it improves a bit the UX, but in practice the vault will only call `onComputeDynamicFee...` and `onAfter...Liquidity` hooks if onRegister was called. Also, these functions are `view`, so there's no risk of exploit there.

Besides, the admin functions will revert anyway (numTokens won't match with token index) if the pool is not properly initialized with this hook.